messaging/activity/FirebaseMessagingActivityGenerator.cs

```diff
@@ -0,0 +1,172 @@
+"    // probably means this intent was fired by tapping on the app icon.",
+"    Bundle extras = intent.getExtras();",
+"    if (extras == null) {{",
+"      return;",
+"    }}",
+"    String from = extras.getString(EXTRA_FROM);",
+"    String messageId = extras.getString(EXTRA_MESSAGE_ID_KEY);",
+"    if (messageId == null) {{",
+"      messageId = extras.getString(EXTRA_MESSAGE_ID_KEY_SERVER);",
+"    }}",
+"    if (from != null && messageId != null) {{",
+"      Intent message = new Intent(this, MessageForwardingService.class);",
+"      message.setAction(MessageForwardingService.ACTION_REMOTE_INTENT);",
+"      message.putExtras(intent);",
+"      message.setData(intent.getData());",
+"      MessageForwardingService.enqueueWork(this, message);",
+"    }}",
+"    setIntent(intent);",
+"  }}",
+"",
+"  /**",
+"   * Dispose of the mUnityPlayer when restarting the app.",
+"   *",
+"   * <p>This ensures that when the app starts up again it does not start with stale data.",
+"   */",
+"  @Override",
+"  protected void onCreate(Bundle savedInstanceState) {{",
+"    if (mUnityPlayer != null) {{",
+"      mUnityPlayer.quit();",
+"      mUnityPlayer = null;",
+"    }}",
+"    super.onCreate(savedInstanceState);",
+"  }}",
+"}}"
+  };
+  private readonly string BaseActivityClass = "UnityPlayerActivity";
+  private readonly string BaseGameActivityClass = "UnityPlayerGameActivity";
+
+  private readonly string GeneratedFileTag = "FirebaseMessagingActivityGenerated";
+  // If this tag is present on the generated file, it will not be replaced.
+  private readonly string PreserveTag = "FirebasePreserve";
+
+  private readonly string OutputPath = Path.Combine("Plugins", "Android");
+  private readonly string OutputFilename = "MessagingUnityPlayerActivity.java";
+
+  public int callbackOrder { get { return 0; } }
+  public void OnPreprocessBuild(BuildReport report) {
+    // Only run this logic when building for Android.
+    if (EditorUserBuildSettings.activeBuildTarget != BuildTarget.Android) {
+      return;
+    }
+
+    // Check if the file has already been generated.
+    string[] oldAssetGuids = AssetDatabase.FindAssets("l:" + GeneratedFileTag);
+    if (oldAssetGuids != null && oldAssetGuids.Length > 0) {
+      if (oldAssetGuids.Length != 1) {
+        Debug.LogWarning("FirebaseMessagingActivityEditor found multiple generated files with the label: " +
+                         GeneratedFileTag + " \nNo changes will be made.");
+        return;
+      }
+      string oldAssetPath = AssetDatabase.GUIDToAssetPath(oldAssetGuids[0]);
+      Object oldAsset = AssetDatabase.LoadMainAssetAtPath(oldAssetPath);
+      // If the generated file has been tagged to be preserved, don't change it.
+      string[] labelList = AssetDatabase.GetLabels(oldAsset);
+      if (labelList.Contains(PreserveTag)) {
+        return;
+      }
+      // Delete the old asset.
+      AssetDatabase.DeleteAsset(oldAssetPath);
```

Comment:
I think we should notify the user about the deletion of the file with a `LogInfo` here and let them know the use of `PreserveTag` in the log in case they are modifying the file themselves.

Here I assume we only delete and regenerate when the file content are different.

Reply:
Done